sdk/storage/Azure.Storage.Blobs/src/PartitionedUploader.cs

```diff
@@ -0,0 +1,449 @@
+{
+    internal class PartitionedUploader
+    {
+        // The client we use to do the actual uploading
+        private readonly BlockBlobClient _client;
+
+        // The maximum number of simultaneous workers
+        private readonly int _maxWorkerCount;
+
+        // A pool of memory we use to partition the stream into blocks
+        private readonly ArrayPool<byte> _arrayPool;
+
+        // The size we use to determine whether to upload as a single PUT BLOB
+        // request or stage as multiple blocks.
+        private readonly long _singleBlockThreshold;
+
+        // The size of each staged block.  If null, we'll change between 4MB
+        // and 8MB depending on the size of the content.
+        private readonly int? _blockSize;
+
+        public PartitionedUploader(BlockBlobClient client, StorageTransferOptions transferOptions, long? singleBlockThreshold = null, ArrayPool<byte> arrayPool = null)
+        {
+            _client = client;
+            _arrayPool = arrayPool ?? ArrayPool<byte>.Shared;
+            _maxWorkerCount =
+                transferOptions.MaximumConcurrency ??
+                Constants.Blob.Block.DefaultConcurrentTransfersCount;
+            _singleBlockThreshold = singleBlockThreshold ?? Constants.Blob.Block.MaxUploadBytes;
+            _blockSize = null;
+            if (transferOptions.MaximumTransferLength != null)
+            {
+                _blockSize = Math.Min(
+                    Constants.Blob.Block.MaxStageBytes,
+                    transferOptions.MaximumTransferLength.Value);
+            }
+        }
+
+        public async Task<Response<BlobContentInfo>> UploadAsync(
+            Stream content,
+            BlobHttpHeaders blobHttpHeaders,
+            IDictionary<string, string> metadata,
+            BlobRequestConditions conditions,
+            IProgress<long> progressHandler,
+            AccessTier? accessTier = default,
+            CancellationToken cancellationToken = default)
+        {
+            // If we can compute the size and it's small enough
+            if (TryGetLength(content, out long length) && length < _singleBlockThreshold)
+            {
+                // Upload it in a single request
+                return await _client.UploadAsync(
+                    content,
+                    blobHttpHeaders,
+                    metadata,
+                    conditions,
+                    accessTier,
+                    progressHandler,
+                    cancellationToken)
+                    .ConfigureAwait(false);
+            }
+
+            // If the caller provided an explicit block size, we'll use it.
+            // Otherwise we'll adjust dynamically based on the size of the
+            // content.
+            int blockSize =
+                _blockSize != null ? _blockSize.Value :
+                length < Constants.LargeUploadThreshold ?
+                    Constants.DefaultBufferSize :
+                    Constants.LargeBufferSize;
```

Comment:
`_blockSize` is null on construction and a new instance of this class is constructed for every method invocation. The existence of both `_blockSize` and local `blockSize` seems to be an unnecessary layer of logic. I'd rather the field `_blockSize` be set correctly in the constructor and used throughout the class.

Reply:
I don't think it's unnecessary.  It's a clean split between the values that control our upload logic (passed into the .ctor/saved as fields) and the values passed to the service (passed to Upload/used as params).  We could jam everything into a single Upload call and thread it all around as parameters, but I don't think it would be cleaner or easier to read.